# AULA 16 - Noções do pacote PANDAS -  parte B
## Algoritmos e Estrutura de Dados I - Prof. Piva
-----------

### Preenchendo Valores Ausentes em DataFrames com Pandas
A função fillna() é utilizada para preencher valores ausentes em um dataframe.
Oferece, para tanto, muitas opções: valores específicos, função agregada (media, moda etc), valor anterior, valor seguinte etc.

In [1]:
# Importando o Pandas
import pandas as pd
# Verificando a Versão do Pandas
#pd.set_options (pat, value)
pd.__version__

'2.2.2'

In [2]:
# Importando um dataset que está em um arquivo CSV
df = pd.read_csv("dataset.csv")

In [12]:
df.head(5)

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [23]:
df.tail()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
9989  CA-2014-110422  2014-01-21   TB-21400  Consumer  United States  South   
9990  CA-2017-121258  2017-02-26   DB-13060  Consumer  United States   West   
9991  CA-2017-121258  2017-02-26   DB-13060  Consumer  United States   West   
9992  CA-2017-121258  2017-02-26   DB-13060  Consumer  United States   West   
9993  CA-2017-119914  2017-05-04   CC-12220  Consumer  United States   West   

           ID_Produto        Categoria  \
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
9989                             Ultra Door Pull Handle       25.248   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...       91.960   
9991                              Aastra 57i VoIP phone      258.576   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"       29.600   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...      243.160   

      Quantidade  
9989         3.0  
9990         2.0  
9991         2.0  
9992         4.0  
9993         2.0

In [25]:
df.shape

(9994, 11)

In [28]:
# Verificando a quantidade de valores ausentes no dataFrame
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [31]:
df.describe()

Valor_Venda   Quantidade
count   9994.000000  9992.000000
mean     229.858001     3.789832
std      623.245101     2.225246
min        0.444000     1.000000
25%       17.280000     2.000000
50%       54.490000     3.000000
75%      209.940000     5.000000
max    22638.480000    14.000000

In [36]:
# Calculando a Moda (moda é uma medida estatística que representa o elemento que mais se repete no conjunto de dados)
# No caso, vamos retornar a moda da coluna 'Quantidade'
# Peço para contar todos os valores com value_counts()
# Em seguida, eu vou retornar a posição 0 (o primeiro elemento) - que é o que mais se repete (maior qtd)

moda = df['Quantidade'].value_counts().head().index[0]

In [37]:
print(moda)

3.0


In [38]:
# Vamos preencher os valores ausentes com esse valor da moda.
# vamos utilizar a função fillna()
df['Quantidade'].fillna(value = moda, inplace = True)

# o parâmetro nomeado inplace é utilizado para Salvar o resultado no mesmo dataframe. 
# Se não utilizarmos o parâmetro como True, ele irá realizar a operação e criar uma cópia do dataframe
# Não alterará o dataframe original. Isso é padrão no Pandas (qualquer operação ele sempre realiza, criando uma cópia do df)

C:\Users\Aluno\AppData\Local\Temp\ipykernel_14896\449297895.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Quantidade'].fillna(value = moda, inplace = True)


In [40]:
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

Esta estratégia de preenchimento chama-se **INTERPOLAÇÃO**. Utilizamos uma medida estatística da coluna para preencher os valores ausentes.

### Consulta de Dados no DataFrame Pandas (query)
Os dataframes criados são tabelas. Como tal, podemos realizar consultas e retornar valores...

In [41]:
# Por exemplo... existe uma coluna: Valor_Venda  ... vamos entendê-la com describe()
df.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

Veja que esse resumo estatístico nos trás informações importantes: valor mínimo: 0.44  Valor máximo: 22638.48
Valor médio: 229,85

In [42]:
# Vamos fazer uma consulta com a função query().
df2 = df.query('229 < Valor_Venda < 10000')

In [43]:
df2.Valor_Venda.describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [44]:
# Uma nova consulta, tomando como base esse segundo dataframe criado a partir da primeira consulta.
df3 = df2.query('Valor_Venda > 766')

In [45]:
df3.Valor_Venda.describe()

count     687.000000
mean     1615.597817
std      1203.403998
min       767.214000
25%       914.200000
50%      1212.960000
75%      1763.575000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [46]:
df3.head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
3   US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
7   CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
10  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
11  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
24  CA-2015-106320  2015-09-25   EB-13870  Consumer  United States   West   

         ID_Produto   Categoria  \
3   FUR-TA-10000577   Furniture   
7   TEC-PH-10002275  Technology   
10  FUR-TA-10001539   Furniture   
11  TEC-PH-10002033  Technology   
24  FUR-TA-10000577   Furniture   

                                     Nome_Produto  Valor_Venda  Quantidade  
3   Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
7                  Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
10       Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11  Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
24  Bretford CR4500 Series Slim Rectangular Table    1044.6300         3.0

### Verificando a Ocorrência de Valores em uma Coluna

No nosso dataset, existe uma coluna 'Quantidade', que representa a quantidade de itens vendidos em cada venda.
Como podemos filtrar os dados, para identificar as vendas que tiveram quantidades vendidas iguais a 5, 9 ou 11 itens?
Uma possibilidade é utilizarmos o método isin() para checar diversos valores em uma coluna (semelhante a cláusula IN em SQL)

In [47]:
#Verificando o shape (formato) do dataframe. Ele retornará o número de linhas e o número de colunas.
df.shape

(9994, 11)

In [48]:
# Fazendo o filtro das quantidades de itens vendidos: 5, 9 e 11
df[ df['Quantidade'].isin([5, 9, 11]) ]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
36    CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
...              ...         ...        ...          ...            ...   
9967  CA-2017-153871  2017-12-11   RB-19435     Consumer  United States   
9971  CA-2015-103772  2015-06-28   MP-17470  Home Office  United States   
9973  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
36    Central  FUR-FU-10003664        Furniture   
...       ...              ...              ...   
9967     East  OFF-BI-10004209  Office Supplies   
9971    South  OFF-AR-10000538  Office Supplies   
9973     West  TEC-PH-10004080       Technology   
9981     East  OFF-LA-10004484  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
36    Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200   
...                                                 ...          ...   
9967            Fellowes Twister Kit, Gray/Clear, 3/pkg      40.2000   
9971  Boston Model 1800 Electric Pencil Sharpener, Gray     140.7500   
9973                           Avaya 5410 Digital phone     271.9600   
9981                                          Avery 476      16.5200   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
9            5.0  
10           9.0  
14           5.0  
36           5.0  
...          ...  
9967         5.0  
9971         5.0  
9973         5.0  
9981         5.0  
9988         5.0  

[1522 rows x 11 columns]

In [49]:
# Podemos obter um shape, depois de um processo de fatiamento. Vamos pegar o exemplo anterior...
df[ df['Quantidade'].isin([5, 9, 11]) ].shape

(1522, 11)

In [50]:
# Podemos aplicar o filtro e retornar apenas uma parte do resultado. Por exemplo... vamor retornar apenas as 10 primeiras linhas
df[ df['Quantidade'].isin([5, 9, 11]) ][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
37  CA-2015-117415  2015-12-27   SN-20710  Home Office  United States   
54  CA-2016-105816  2016-12-11   JM-15265    Corporate  United States   
56  CA-2016-111682  2016-06-17   TB-21055     Consumer  United States   
57  CA-2016-111682  2016-06-17   TB-21055     Consumer  United States   
66  US-2015-164175  2015-04-30   PS-18970  Home Office  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
36  Central  FUR-FU-10003664        Furniture   
37  Central  OFF-EN-10002986  Office Supplies   
54     East  TEC-PH-10002447       Technology   
56     East  OFF-PA-10001569  Office Supplies   
57     East  FUR-CH-10003968        Furniture   
66  Central  FUR-CH-10001146        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0  
37  #10-4 1/8" x 9 1/2" Premium Diagonal Seam Enve...     113.3280         9.0  
54                   AT&T CL83451 4-Handset Telephone    1029.9500         5.0  
56                                          Xerox 232      32.4000         5.0  
57                           Novimex Turbo Task Chair     319.4100         5.0  
66        Global Value Mid-Back Manager's Chair, Gray     213.1150         5.0

### Trabalhando com Operadores Lógicos p/ Manipulação de Dados com Pandas
AND, OR, NOT...

In [54]:
# Vamos fazer um filtro das vendas que ocorreram para o segmento "Home Office" E na região "South"
df[ (df.Segmento == 'Home Office') & (df.Regiao == 'South') ].head()

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais Regiao  \
182  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
183  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
184  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
231  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   
232  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   

          ID_Produto   Categoria  \
182  TEC-PH-10003273  Technology   
183  TEC-PH-10004896  Technology   
184  TEC-AC-10002345  Technology   
231  FUR-TA-10001705   Furniture   
232  FUR-TA-10003473   Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
182                                      AT&T TR1909W     503.9600         4.0  
183                        Nokia Lumia 521 (T-Mobile)     149.9500         5.0  
184                 HP Standard 104 key PS/2 Keyboard      29.0000         2.0  
231  Bush Advantage Collection Round Conference Table     233.8600         2.0  
232        Bretford Rectangular Conference Table Tops     620.6145         3.0

In [55]:
# Vamos fazer um filtro das vendas que ocorreram para o segmento "Home Office" OU na região "South"
df[ (df.Segmento == 'Home Office') | (df.Regiao == 'South') ].tail()

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9980  US-2015-151435  2015-09-06   SW-20455     Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9989  CA-2014-110422  2014-01-21   TB-21400     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
9979   West  OFF-BI-10002026  Office Supplies   
9980  South  FUR-TA-10001029        Furniture   
9987  South  TEC-AC-10001539       Technology   
9988  South  TEC-PH-10004006       Technology   
9989  South  FUR-FU-10001889        Furniture   

                                           Nome_Produto  Valor_Venda  \
9979                  Ibico Recycled Linen-Style Covers      437.472   
9980                         KI Adjustable-Height Table       85.980   
9987  Logitech G430 Surround Sound Gaming Headset wi...       79.990   
9988                    Panasonic KX - TS880B Telephone      206.100   
9989                             Ultra Door Pull Handle       25.248   

      Quantidade  
9979        14.0  
9980         1.0  
9987         1.0  
9988         5.0  
9989         3.0

In [57]:
# Vamos fazer um filtro das vendas que NÃO ocorreram para o segmento "Home Office" E NÃO foram na região "South"
df[ (df.Segmento != 'Home Office') & (df.Regiao != 'South') ].sample(5)
# sample() retorna uma amostra

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais   Regiao  \
673   CA-2017-130351  2017-12-05   RB-19570  Consumer  United States  Central   
5239  CA-2016-116561  2016-09-11   EB-14110  Consumer  United States     West   
1940  CA-2016-112109  2016-07-08   JE-15715  Consumer  United States     West   
9151  US-2015-106873  2015-09-24   KM-16720  Consumer  United States     West   
7705  CA-2016-114601  2016-08-26   AA-10480  Consumer  United States  Central   

           ID_Produto        Categoria  \
673   OFF-AP-10004532  Office Supplies   
5239  OFF-ST-10004186  Office Supplies   
1940  FUR-CH-10004287        Furniture   
9151  OFF-AR-10003179  Office Supplies   
7705  TEC-AC-10003911       Technology   

                                           Nome_Produto  Valor_Venda  \
673   Kensington 6 Outlet Guardian Standard Surge Pr...       61.440   
5239  Stur-D-Stor Shelving, Vertical 5-Shelf: 72"H x...      332.940   
1940                           SAFCO Arco Folding Chair      662.880   
9151        Dixon Ticonderoga Core-Lock Colored Pencils       14.576   
7705  NETGEAR AC1750 Dual Band Gigabit Smart WiFi Ro...      479.970   

      Quantidade  
673          3.0  
5239         3.0  
1940         3.0  
9151         2.0  
7705         3.0

### Agrupamento de Dados em DataFrames com GroupBy

Podemos utilizar a função groupby() do Pandas para explorar e revelar relacionamento entre variáveis em conjunto de dados.

Vamos fazer a combinação dos dados por **Segmento / Região / Valor_Venda** e calcular a _taxa média_ de vendas de cada grupo.

In [59]:
df[ ['Segmento', 'Regiao', 'Valor_Venda'] ].groupby( ['Segmento', 'Regiao'] ).mean()

Valor_Venda
Segmento    Regiao              
Consumer    Central   207.946728
            East      238.875539
            South     233.390180
            West      217.033955
Corporate   Central   234.763466
            East      228.516929
            South     238.992025
            West      235.265911
Home Office Central   208.248046
            East      253.911805
            South     272.996329
            West      239.442692

### Agregação Múltipla com GroupBy
Vamos utilizar agora a função agg() para realizar a agregação múltipla.

Para tanto, vamos filtrar os dados, extraindo 3 colulas: Segmento, Região e Valor_Venda.
Em seguida, vamos agrupar por duas colunas: Segmento e Região.
Por fim, vamos agregar os dados, calculando média, desvio padrão e contagem de elementos, para a coluna que ficou fora do group by (Valor_Venda).

In [ ]:
df[ ['Segmento', 'Regiao', 'Valor_Venda'] ].groupby( ['Segmento', 'Regiao'] ).agg(['mean', 'std', 'count'])

### Filtrando DataFrames com elementos Strings
Pandas oferece diversas funções para manipulação de strings.
As funções startswith() e endswith() são muito utilizadas para filtragem de strings

In [60]:
df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [61]:
# Filtro dos elementos que iniciam com os caractes: 'Con'
df[ df.Segmento.str.startswith('Con') ].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0

In [62]:
# Filtro dos elementos que terminam com os caractes: 'mar'
df[ df.Segmento.str.endswith('mer') ].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0